Documentation/manpages/s3d_x11gate.1
.TH "s3d_x11gate" "1" 
.SH "NAME" 
s3d_x11gate \(em  3d gateway for x11 desktops  
.SH "SYNOPSIS" 
.PP 
\fBs3d_x11gate\fR [\-h|\-\-help]  [\-\-s3d-url \fIurl\fR]  
.SH "DESCRIPTION" 
.PP 
\fBs3d_x11gate(1)\fR is a 3D gateway for x11 desktops. 
 
.PP 
 
.SH "OPTIONS" 
.PP 
This program follows the usual GNU command line syntax. A summary of 
the options supported by \fBs3d_x11gate\fR is below. 
 
.IP "\fB-h\fP, \fB\-\-help\fP 				" 10 
Show summary of options and exit.  
.IP "\fB\-\-s3d-url \fIurl\fR\fP" 10 
skip S3D environment and connect to this url  
.SH "SEE ALSO" 
.PP 
\fBs3d\fR(1). 
.PP 
\fI s3d project page on berlios  (link to URL http://s3d.berlios.de/) \fR  
.\" created by instant / docbook-to-man, Fri 29 Aug 2008, 18:39 


ChangeLog
(detailed ChangeLog can be generated with "svn log" in a subversion
Checkout. We will only give a brief summary of the changes here.)

2008-08-27 Simon Wunderlich <dotslash@packetmixer.de>
	- GLUT specific parts were removed from server
	- server will generate mipmaps for textures
	- SHM connected server and clients will now share their texture buffers
	  over shm. Clients will only send update notices to inform the
	  server
	- s3dosm is now OSM api 0.5 compatible
	- C++ compatible api headers (no abi changes)
	- of course: fixed various bugs (Thanks Sven for your effort!!)
2007-10-24 Simon Wunderlich - <dotslash@packetmixer.de>
	- Released S3D Version 0.2
	- now using cmake instead of autohell, which seems to be faster,
	  more configurable and at least comprehensible (Thanks to Sven
	  Eckelmann)
	- meshs3d is a complete rewrite of olsrs3d, using faster
	  data structures and clean design (implemented by Andreas Langer)
	- s3dosm is a new program to navigate on openstreetmap maps.
	  it has gps-support, superseding the old s3dgps which operated
	  on bitmap map files.
	- kism3d (implemented by Marek Lindner) visualizes the wireless
	  networks and works as a frontend to kismet
	- texture support is now using mipmaps
	- added a few widgets (textbox/scrollbars)
	- of course: fixed various bugs
2006-09-19 Simon Wunderlich - <dotslash@packetmixer.de>
	- Released S3D Version 0.1.1
	- client is now libs3d
	- added widget library libs3dw
	- olsrs3d uses widgets for search input
2006-03-19  Simon Wunderlich - <dotslash@packetmixer.de>
	- FINALLY, PROUDLY PRESENTING: S3D VERSION 0.1!!!
	- this changelog is pretty incomplete. Please use the
	  subversion ChangeLog (by typing svn log) for updates.
	- We will stop this Changelog now, as svn does the dirty work :)
2006-03-15  Andreas Kohn - <andreas@syndrom23.de>
	- some configure.ac changes to fix glut/sdl detection code
	- some autoconf magic to fix pthread detection for esoteric arch's.
	- a big hammer to prevent fortran checks
2006-03-14  Simon Wunderlich - <dotslash@packetmixer.de>
	- added push/pep/pop-line and pep-vertex
	- dot_mcp now has egoshooter-mode (w,a,s,d)
	- added some Documentation entries
2006-03-10  Simon Wunderlich - <dotslash@packetmixer.de>
	- added rc functionality: use /etc/s3drc or ~/.s3drc to
	  automatically fire up your mcp or anything you want.
2006-03-07  Simon Wunderlich - <dotslash@packetmixer.de>
	- fixed -O2 problem
2006-03-03  Simon Wunderlich - <dotslash@packetmixer.de>
	- using values from config.h, user does not have to edit server/config.h
	  or clients/config.h anymore to fit his/her needs
	- Finally, Font triangulating is stable (at least it seems so ;)
2006-02-27  Simon Wunderlich - <dotslash@packetmixer.de>
	- added GPL/LGPL copyright notes (phew -.-)
	- implemented object search path as : delimited list
	- INFINITY define into #ifndef
	- splitting README and INSTALL
	- security check for strncat and strncpy functions
2006-02-22  Simon Wunderlich - <dotslash@packetmixer.de>
	- only select the nearest hit when clicking, not all on the line
	- fopen errors are not that important when looking through the
	  search path ...
	- SIGNAL -> WITH_SIGNALS
2006-02-20  Simon Wunderlich - <dotslash@packetmixer.de>
	- added mplayer patch: patch your mplayer, use -vo s3d on
	  the commandline and watch your videos in s3d in 3d space - yay
2006-02-18	Simon Wunderlich - <dotslash@packetmixer.de>	
	- by commenting out DEBUG define, all debug messages
	  will not even be [call]ed
	- dot_mcp added view reset button, fixed rotate button 
	- dot_mcp has now some "GO!!" menu
	- 3dsread.c x-axis inversion fixed
	- fixed 0.0 size application problem
	- removed c++ styles in olsrs3d
2006-02-17	Simon Wunderlich - <dotslash@packetmixer.de>	
	- gintl() and gints() added in 3dsread.c, hopefully
	  improving endianess compatibility
	- moved dot_mcp to apps
	- abadoning my own tries and the glu-tesselator 
	  for triangulating the fonts and using public
	  domain code for seidels algorithm by Narkhede A. and
	  Manocha D.
2006-01-30	Andreas Kohn - <andreas@syndrom23.de>
	- rework larger parts of the build infrastructure, 
	  use automake and libtool.
2005-22-12	Simon Wunderlich - <dotslash@packetmixer.de>
	- fixed: server did not unfocus when app quits ...
	- mcp got some nice rotation mode
2005-21-12	Simon Wunderlich - <dotslash@packetmixer.de>
	- fixed some focusing/camera update issues
	- perfomance improvement in olsrs3d
2005-16-12	Simon Wunderlich - <dotslash@packetmixer.de>
	- telling the mcp/programs aspect ratio of the cam
	- new layout in the mcp.
2005-12-12	Simon Wunderlich - <dotslash@packetmixer.de>
	- tuning on the configure/make system
	- added glutInit(), to make glut happy
2005-05-12	Simon Wunderlich - <dotslash@packetmixer.de>
	- some more work on the olsr visualizer, testing with 
	  more graphs, hope it won't explode anymore ;)
2005-27-11	Simon Wunderlich - <dotslash@packetmixer.de>
	- finished olsrs3d, the olsr-visualizer
2005-26-11	Simon Wunderlich - <dotslash@packetmixer.de>
	- removed some race condition bug in the server
2005-25-11	Simon Wunderlich - <dotslash@packetmixer.de>
	- starting olsrs3d
2005-24-10	Simon Wunderlich - <dotslash@packetmixer.de>
	- some more bugfixes in the autoconf-script.
2005-23-10	Simon Wunderlich - <dotslash@packetmixer.de>
	- autoconf script nearly completed, it only compiled
	  the apps you have the right libraries available.
2005-14-10	Simon Wunderlich - <dotslash@packetmixer.de>
	- so long no update ... project is kind of sleeping :(
	- autoconf configure script started. Now you can check
	  for your libs before compiling.
2005-10-7	Simon Wunderlich - <dotslash@packetmixer.de>
	- filebrowser is now some cake with pies to select :)
2005-04-7	Simon Wunderlich - <dotslash@packetmixer.de>
	- massive cleanup in s3dvt, removing nextchar in 
	  linestructures-weirdness, just having cx/cy
	  as cursor positions.
2005-03-7	Simon Wunderlich - <dotslash@packetmixer.de>

apps/s3d_x11gate/s3d_x11gate.c
/*
 * s3d_x11gate.c
 *
 * Copyright (C) 2004-2008 Simon Wunderlich <dotslash@packetmixer.de>
 *
 * This file is part of s3d_x11gate, a 3d gateway for x11 desktops.
 * See http://s3d.berlios.de/ for more updates.
 *
 * s3d_x11gate is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * s3d_x11gate is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with s3d_x11gate; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */


#include <s3d.h>   /*  s3d_*() */
#include <stdlib.h>   /*  getenv() */
#include <stdio.h>   /*  printf() */
#include <X11/Xlib.h>  /*  Ximage, Display, X*() */
#include <X11/Xutil.h>  /*  XDestroyImage() */
#define XK_MISCELLANY
#include <X11/keysymdef.h>  /* keysyms */
#include <X11/extensions/XTest.h> /* keyboard/mouse input via s3d */
#include <X11/extensions/XShm.h> /* */
#include <time.h>  /* nanosleep() */
#include <sys/time.h>  /* gettimeofday */
#include <sys/ipc.h>
#include <sys/shm.h>
#include <config-s3d.h>

#ifndef S3DX11UNUSED
#if defined(UNUSEDPARAM_ATTRIBUTE)
#define S3DX11UNUSED(x) (x)__attribute__((unused))
#elif defined(UNUSEDPARAM_OMIT)
#define S3DX11UNUSED(x) /* x */
#else
#define S3DX11UNUSED(x) x
#endif
#endif

static int oid;
static XImage *image;
static Display *dpy = NULL;
static int window, scr;
static unsigned int width, height, depth;
static Visual *visual;
static XShmSegmentInfo shminfo;
static char *tex_image = NULL, *otex_image = NULL, *img1, *img2;



static struct timeval start, end;
static int iterations;
static float count[3];

static int get_shift(uint32_t t)
{
	int i = 0;
	while (t) {
		t >>= 1;
		i++;
	}
	return(i);
}
static void mainloop(void)
{
#define MAGIC_CHANGED ((unsigned int)~0)
	unsigned int x, y;
	int rs, gs, bs;
	uint32_t d;
	int bpp;
	char *swap_timg;
	unsigned int last_change, start_change;
	gettimeofday(&end, NULL);
	count[0] += (end.tv_sec - start.tv_sec) * 10000000 + end.tv_usec - start.tv_usec;
	start.tv_sec = end.tv_sec;
	start.tv_usec  = end.tv_usec;

	/* image = XGetImage(dpy,window,0,0,width,height,AllPlanes,ZPixmap);*/
	XShmGetImage(dpy, window, image, 0, 0, 0xffffffff);
	gettimeofday(&end, NULL);
	count[1] += (end.tv_sec - start.tv_sec) * 10000000 + end.tv_usec - start.tv_usec;
	start.tv_sec = end.tv_sec;
	start.tv_usec  = end.tv_usec;
	if (image->format == ZPixmap) {
		printf("Ximage: %dx%d, format %d (%d), bpp: %d, depth %d, pad %d\n", image->width, image->height, image->format, ZPixmap,
		       image->bits_per_pixel, image->depth, image->bitmap_pad);
		rs = get_shift(image->red_mask) - 8;
		gs = get_shift(image->green_mask) - 8;
		bs = get_shift(image->blue_mask) - 8;

		bpp = (image->bits_per_pixel / 8);
		/* rgb is not bgr */
		rs = rs;
		gs = gs - 8;
		bs = bs - 16;
		printf("Ximage: rgb: %d|%d|%d\n", rs, gs, bs);;
		/*  printf("red: size %d, offset %d\n",rs,roff);
		  printf("green: size %d, offset %d\n",gs,goff);
		  printf("blue: size %d, offset %d\n",bs,boff);
		  printf("bits per pixel:%d\n",bpp);*/
		last_change = MAGIC_CHANGED;
		start_change = MAGIC_CHANGED;
		for (y = 0;y < height;y++) {

			for (x = 0;x < width;x++) {
				d = *((uint32_t *)(image->data + (y * width + x) * bpp));
				((uint32_t *)tex_image)[(y*width+x)] =
				        (rs > 0 ? ((d & image->red_mask) >> rs) : ((d & image->red_mask) << -rs)) |
				        (gs > 0 ? ((d & image->green_mask) >> gs) : ((d & image->green_mask) << -gs)) |
				        (bs > 0 ? ((d & image->blue_mask) >> bs) : ((d & image->blue_mask) << -bs)) |
				        255u << 24;
				if (((uint32_t *)tex_image)[(y*width+x)] !=
				                ((uint32_t *)otex_image)[(y*width+x)])
					last_change = y;
			}
			if (last_change != MAGIC_CHANGED) {
				if (start_change == MAGIC_CHANGED) {
					start_change = y;
					/*      printf("setting start_change to %d\n",start_change); */
				}
				if (last_change != y) {  /*  last change is already over, post it! */
					s3d_load_texture(oid, 0, 0, start_change, width, last_change - start_change + 1, (unsigned char *)tex_image + start_change*width * sizeof(uint32_t));
					start_change = MAGIC_CHANGED;
					last_change = MAGIC_CHANGED;
				}
			}
		}
		/*  posting the last bit, maybe */
		if (last_change != MAGIC_CHANGED) {
			/*   printf("last one: [%d-%d]",start_change,last_change);*/
			s3d_load_texture(oid, 0, 0, start_change, width, last_change - start_change, (unsigned char *)tex_image + start_change*width * sizeof(uint32_t));
		}
		/*   s3d_load_texture(oid,0,0,0,width,height,tex_image); */
		/*  swap images */
		swap_timg = tex_image;
		tex_image = otex_image;
		otex_image = swap_timg;
	}
	gettimeofday(&end, NULL);
	count[2] += (end.tv_sec - start.tv_sec) * 10000000 + end.tv_usec - start.tv_usec;
	start.tv_sec = end.tv_sec;
	start.tv_usec  = end.tv_usec;
	iterations++;
	/* XDestroyImage(image);*/
	/* nanosleep(&t,NULL); */
}
static int keypress(struct s3d_evt *event)
{
	int key;
	int kc;
	key = *((unsigned short *)event->buf);
	printf("received key %d ", key);
	kc = XKeysymToKeycode(dpy, key);
	if (kc == 0) {
		kc = XKeysymToKeycode(dpy, 0xFF00 + key);
		printf(" (%04x) ", 0xFF00 + key);
	}
	if (kc == 0) {
		if (key == 8) {
			kc = 22;
			printf("!backspace!");
		}
	}
	printf("using key: %d, keycode %d (%04x)\n", key, kc, kc);
	if (kc != 0)
		XTestFakeKeyEvent(dpy, kc, 1, 1);
	/*     XTestFakeKeyEvent(dpy, kc, 0, 1);*/
	return(0);

}
static int mouseclick(struct s3d_evt *S3DX11UNUSED(event))
{
	int i;
	printf("thats it, collecting:\n");
	for (i = 0;i < 3;i++)
		printf("[%d] %f\n", i, count[i] / iterations);
	exit(0);
}
int main(int argc, char **argv)
{
	const char *disp = NULL;
	int a, b, c, d;
	int xt;
	if (disp == NULL) disp = getenv("DISPLAY");
	if (disp == NULL) disp = "";  /*  fallback */
	dpy = XOpenDisplay(disp);
	if (!dpy) {
		printf("couldn't open display\n");
		return(-1);
	}
	count[0] = count[1] = count[2] = 0;
	iterations = 0;
	if (!s3d_init(&argc, &argv, "X11-gate")) {
		scr = DefaultScreen(dpy);
		window = RootWindow(dpy, scr);
		width = DisplayWidth(dpy, scr);
		height = DisplayHeight(dpy, scr);
		visual = DefaultVisual(dpy, scr);
		depth = DefaultDepth(dpy, scr);
		XLockDisplay(dpy);
		xt = XTestQueryExtension(dpy, &a, &b, &c, &d);
		XUnlockDisplay(dpy);
		if (xt) {
			printf("having xtest extension ...\n");
		}
		/* X11 shm - http://www.xfree86.org/current/mit-shm.html */

		image = XShmCreateImage(dpy, visual, depth, ZPixmap, NULL, &shminfo, width, height);
		shminfo.shmid = shmget(IPC_PRIVATE, image->bytes_per_line * image->height, IPC_CREAT | 0777);
		shminfo.shmaddr = image->data = (char*)shmat(shminfo.shmid, NULL, 0);
		shmctl(shminfo.shmid, IPC_RMID, NULL);
		shminfo.readOnly = False;
		if (!XShmAttach(dpy, &shminfo))
			printf("cannot use the shared memory segment .. :( \n");
		else
			printf("can use share segment :D\n");
		XSync(dpy, False);

		s3d_set_callback(S3D_EVENT_OBJ_CLICK, mouseclick);
		s3d_set_callback(S3D_EVENT_KEY, keypress);
		printf("screen: %dx%d\n", width, height);
		img1 = (char*)malloc(width * height * sizeof(uint32_t));
		img2 = (char*)malloc(width * height * sizeof(uint32_t));
		tex_image = img1;
		otex_image = img2;
		oid = s3d_new_object();
		s3d_push_vertex(oid, -5, -5, 0);
		s3d_push_vertex(oid, 5, -5, 0);
		s3d_push_vertex(oid, 5, 5, 0);
		s3d_push_vertex(oid, -5, 5, 0);
		s3d_push_material_a(oid,
		                    0.8, 0.0, 0.0 , 1.0,
		                    1.0, 1.0, 1.0 , 1.0,
		                    0.8, 0.0, 0.0 , 1.0);
		s3d_push_polygon(oid, 0, 2, 1, 0);
		s3d_pep_polygon_tex_coord(oid, 0.0, 1.0,
		                          1.0, 0.0,
		                          1.0, 1.0);
		s3d_push_polygon(oid, 0, 3, 2, 0);
		s3d_pep_polygon_tex_coord(oid, 0.0, 1.0,
		                          0.0, 0.0,
		                          1.0, 0.0);
		s3d_push_texture(oid, width, height);
		/*  push data on texture 0 position (0,0) */
		s3d_pep_material_texture(oid, 0); /*  assign texture 0 to material 0 */
		s3d_flags_on(oid, S3D_OF_VISIBLE | S3D_OF_SELECTABLE);
		gettimeofday(&start, NULL);
		s3d_mainloop(mainloop);
		free(img1);
		free(img2);
	}
	s3d_quit();
	return(0);
}
